File: server/src/services/rate_limit.py
"""Per-IP daily cap on POST /sessions (local direct API only)."""

from __future__ import annotations

import os
import threading
from datetime import UTC, datetime

from fastapi import Request

# Keep in sync with worker/src/rate_limit.ts
RATE_LIMIT_DETAIL = "今日创建会话次数已达上限，请明天再试。"


class RateLimitError(Exception):
    def __init__(self, detail: str = RATE_LIMIT_DETAIL) -> None:
        self.detail = detail
        super().__init__(detail)


_lock = threading.Lock()
_buckets: dict[str, tuple[str, int]] = {}


def _day_key() -> str:
    return datetime.now(UTC).strftime("%Y-%m-%d")


def _limit() -> int:
    return int(os.getenv("RATE_LIMIT_SESSIONS_PER_DAY", "20"))


def should_apply_rate_limit(request: Request) -> bool:
    """Skip when disabled or request already passed Worker edge limit."""
    flag = os.getenv("RATE_LIMIT_ENABLED", "1").strip().lower()
    if flag in ("0", "false", "no"):
        return False
    if request.headers.get("cf-connecting-ip"):
        return False
    return True


def client_ip(request: Request) -> str:
    forwarded = request.headers.get("x-forwarded-for")
    if forwarded:
        return forwarded.split(",")[0].strip()
    if request.client and request.client.host:
        return request.client.host
    return "unknown"


def check_session_create_rate_limit(client_key: str) -> None:
    limit = _limit()
    if limit <= 0:
        return
    day = _day_key()
    with _lock:
        prev_day, count = _buckets.get(client_key, (day, 0))
        if prev_day != day:
            count = 0
        if count >= limit:
            raise RateLimitError()
        _buckets[client_key] = (day, count + 1)


def reset_for_tests() -> None:
    with _lock:
        _buckets.clear()

File: worker/src/rate_limit.ts
/** Daily cap on POST /api/sessions (edge). Keep detail in sync with server rate_limit.py */

export const RATE_LIMIT_DETAIL =
  "今日创建会话次数已达上限，请明天再试。";

type Bucket = { day: string; count: number };

const buckets = new Map<string, Bucket>();

function dayKey(): string {
  return new Date().toISOString().slice(0, 10);
}

export function clientIp(request: Request): string {
  const cf = request.headers.get("CF-Connecting-IP");
  if (cf?.trim()) return cf.trim();
  const xff = request.headers.get("X-Forwarded-For");
  if (xff) {
    const first = xff.split(",")[0]?.trim();
    if (first) return first;
  }
  return "unknown";
}

/** Returns a 429 Response when limited; otherwise records this attempt and returns null. */
export function limitSessionCreate(
  request: Request,
  limit: number,
): Response | null {
  if (!Number.isFinite(limit) || limit <= 0) return null;

  const ip = clientIp(request);
  const day = dayKey();
  const key = `${ip}:${day}`;
  const entry = buckets.get(key);
  const count = entry?.day === day ? entry.count : 0;

  if (count >= limit) {
    return new Response(JSON.stringify({ detail: RATE_LIMIT_DETAIL }), {
      status: 429,
      headers: { "Content-Type": "application/json" },
    });
  }

  buckets.set(key, { day, count: count + 1 });
  return null;
}

/** @internal test helper */
export function _resetRateLimitBucketsForTests(): void {
  buckets.clear();
}
